The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16365816175/job/46242706290) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46242706290) [(plain)](https://github.com/rust-lang/rust/commit/2390abed22be00a55a4c9326d09c1a9bd0c47d4e/checks/46242706290/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

+ error[E0449]: visibility qualifiers are not permitted here
+   --> $DIR/pub-priv1.rs:49:9
+    |
+ LL |         pub other_field: PubType,
+    |         ^^^ help: remove the qualifier
+    |
+    = note: enum variants and their fields always share the visibility of the enum they are in
+ 
1 error: crate `priv_dep` from private dependency 'priv_dep' is re-exported
2   --> $DIR/pub-priv1.rs:12:1
3    |

64 LL | pub use priv_dep::OtherType as Renamed;
65    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
66 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:29:5
-    |
- LL |     pub field: OtherType,
-    |     ^^^^^^^^^^^^^^^^^^^^
+ error: aborting due to 11 previous errors
72 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:36:5
-    |
- LL |     pub OtherType,
-    |     ^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:44:21
-    |
- LL |     ActualOtherType(OtherType, PubType),
-    |                     ^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:47:9
-    |
- LL |         field: OtherType,
-    |         ^^^^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:54:1
-    |
- LL | pub type ReexportedPublicGeneric = PublicGenericType<OtherType, ()>;
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:56:1
-    |
- LL | pub type ReexportedPrivateGeneric = PublicGenericType<(), OtherType>;
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: trait `OtherTrait` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:59:1
-    |
- LL | pub struct PublicGenericBoundedType<T: OtherTrait>(T);
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:63:5
-    |
- LL |     pub fn pub_fn_param(param: OtherType) {}
-    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:66:5
-    |
- LL |     pub fn pub_fn_return() -> OtherType { OtherType }
-    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: trait `OtherTrait` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:73:5
-    |
- LL |     type Foo: OtherTrait;
-    |     ^^^^^^^^^^^^^^^^^^^^
- 
- error: trait `OtherTrait` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:76:33
-    |
- LL |     fn required_impl_trait() -> impl OtherTrait;
-    |                                 ^^^^^^^^^^^^^^^
- 
- error: trait `OtherTrait` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:79:33
-    |
- LL |     fn provided_impl_trait() -> impl OtherTrait { OtherType }
-    |                                 ^^^^^^^^^^^^^^^
- 
- error: trait `OtherTrait` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:79:33
-    |
- LL |     fn provided_impl_trait() -> impl OtherTrait { OtherType }
-    |                                 ^^^^^^^^^^^^^^^
-    |
-    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:83:5
-    |
- LL |     fn required_concrete() -> OtherType;
-    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:86:5
-    |
- LL |     fn provided_concrete() -> OtherType { OtherType }
-    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: trait `OtherTrait` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:90:1
-    |
- LL | pub trait WithSuperTrait: OtherTrait {}
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:99:5
-    |
- LL |     type X = OtherType;
-    |     ^^^^^^
- 
- error: trait `OtherTrait` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:103:1
-    |
- LL | pub fn in_bounds<T: OtherTrait>(x: T) { unimplemented!() }
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: trait `OtherTrait` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:106:1
-    |
- LL | pub fn private_return_impl_trait() -> impl OtherTrait { OtherType }
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:109:1
-    |
- LL | pub fn private_return() -> OtherType { OtherType }
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:112:1
-    |
- LL | pub fn private_in_generic() -> std::num::Saturating<OtherType> { unimplemented!() }
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:115:1
-    |
- LL | pub static STATIC: OtherType = OtherType;
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:118:1
-    |
- LL | pub const CONST: OtherType = OtherType;
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:121:1
-    |
- LL | pub type Alias = OtherType;
-    | ^^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:124:1
-    |
- LL | pub type AliasOfAlias = priv_dep::PubPub;
-    | ^^^^^^^^^^^^^^^^^^^^^
- 
- error: trait `OtherTrait` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:129:1
-    |
- LL | impl OtherTrait for PublicWithPrivateImpl {}
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:134:1
-    |
- LL | impl PubTraitOnPrivate for OtherType {}
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:134:1
-    |
- LL | impl PubTraitOnPrivate for OtherType {}
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-    |
-    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:140:1
-    |
- LL | impl From<OtherType> for PublicWithStdImpl {
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:142:5
-    |
- LL |     fn from(val: OtherType) -> Self { Self }
-    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:146:1
-    |
- LL | impl From<PublicWithStdImpl> for OtherType {
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:146:1
-    |
- LL | impl From<PublicWithStdImpl> for OtherType {
-    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-    |
-    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:149:5
-    |
- LL |     fn from(val: PublicWithStdImpl) -> Self { Self }
-    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: type `OtherType` from private dependency 'priv_dep' in public interface
-   --> $DIR/pub-priv1.rs:149:5
-    |
- LL |     fn from(val: PublicWithStdImpl) -> Self { Self }
-    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-    |
-    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
- 
- error: aborting due to 44 previous errors
- 
+ For more information about this error, try `rustc --explain E0449`.
281 

Note: some mismatched output was normalized before being compared
-   --> /checkout/tests/ui/privacy/pub-priv-dep/pub-priv1.rs:49:9
+ error[E0449]: visibility qualifiers are not permitted here
+   --> $DIR/pub-priv1.rs:49:9
+    |
+ LL |         pub other_field: PubType,
+    |         ^^^ help: remove the qualifier
+    |
+    = note: enum variants and their fields always share the visibility of the enum they are in
+ 
+ error: aborting due to 11 previous errors
+ For more information about this error, try `rustc --explain E0449`.
---
To only update this specific test, also pass `--test-args privacy/pub-priv-dep/pub-priv1.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/privacy/pub-priv-dep/pub-priv1.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/privacy/pub-priv-dep/pub-priv1" "-A" "unused" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/privacy/pub-priv-dep/pub-priv1/auxiliary" "-Zunstable-options" "--extern" "priv:priv_dep=/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/privacy/pub-priv-dep/pub-priv1/auxiliary/libpriv_dep.so" "--extern" "priv:pm=/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/privacy/pub-priv-dep/pub-priv1/auxiliary/libpm.so"
stdout: none
--- stderr -------------------------------
error[E0449]: visibility qualifiers are not permitted here
##[error]  --> /checkout/tests/ui/privacy/pub-priv-dep/pub-priv1.rs:49:9
   |
LL |         pub other_field: PubType,
   |         ^^^ help: remove the qualifier
   |
   = note: enum variants and their fields always share the visibility of the enum they are in

error: crate `priv_dep` from private dependency 'priv_dep' is re-exported
##[error]  --> /checkout/tests/ui/privacy/pub-priv-dep/pub-priv1.rs:12:1
   |
LL | pub extern crate priv_dep;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: the lint level is defined here
  --> /checkout/tests/ui/privacy/pub-priv-dep/pub-priv1.rs:9:9
   |
LL | #![deny(exported_private_dependencies)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: macro `m` from private dependency 'priv_dep' is re-exported
##[error]  --> /checkout/tests/ui/privacy/pub-priv-dep/pub-priv1.rs:159:9
   |
LL | pub use priv_dep::m;
   |         ^^^^^^^^^^^

error: macro `fn_like` from private dependency 'pm' is re-exported
##[error]  --> /checkout/tests/ui/privacy/pub-priv-dep/pub-priv1.rs:161:9
   |
LL | pub use pm::fn_like;
   |         ^^^^^^^^^^^

error: derive macro `PmDerive` from private dependency 'pm' is re-exported
##[error]  --> /checkout/tests/ui/privacy/pub-priv-dep/pub-priv1.rs:163:9
   |
LL | pub use pm::PmDerive;
   |         ^^^^^^^^^^^^

error: attribute macro `pm_attr` from private dependency 'pm' is re-exported
##[error]  --> /checkout/tests/ui/privacy/pub-priv-dep/pub-priv1.rs:165:9
   |
LL | pub use pm::pm_attr;
   |         ^^^^^^^^^^^

error: variant `V1` from private dependency 'priv_dep' is re-exported
##[error]  --> /checkout/tests/ui/privacy/pub-priv-dep/pub-priv1.rs:168:9
   |
LL | pub use priv_dep::E::V1;
   |         ^^^^^^^^^^^^^^^

error: type alias `Unit` from private dependency 'priv_dep' is re-exported
##[error]  --> /checkout/tests/ui/privacy/pub-priv-dep/pub-priv1.rs:171:9
   |
LL | pub use priv_dep::Unit;
   |         ^^^^^^^^^^^^^^

error: type alias `PubPub` from private dependency 'priv_dep' is re-exported
##[error]  --> /checkout/tests/ui/privacy/pub-priv-dep/pub-priv1.rs:173:9
   |
LL | pub use priv_dep::PubPub;
   |         ^^^^^^^^^^^^^^^^

error: type alias `PubPriv` from private dependency 'priv_dep' is re-exported
##[error]  --> /checkout/tests/ui/privacy/pub-priv-dep/pub-priv1.rs:175:9
   |
LL | pub use priv_dep::PubPriv;
   |         ^^^^^^^^^^^^^^^^^

error: struct `Renamed` from private dependency 'priv_dep' is re-exported
##[error]  --> /checkout/tests/ui/privacy/pub-priv-dep/pub-priv1.rs:177:9
   |
LL | pub use priv_dep::OtherType as Renamed;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

</details>

--- a/tests/ui/privacy/pub-priv-dep/auxiliary/priv_dep.rs
+++ b/tests/ui/privacy/pub-priv-dep/auxiliary/priv_dep.rs
@@ -10,3 +10,9 @@ macro_rules! m {
 pub enum E {
     V1
 }
+
+struct PrivType;
+
+pub type Unit = ();
+pub type PubPub = OtherType;
+pub type PubPriv = PrivType;

--- a/tests/ui/privacy/pub-priv-dep/pub-priv1.rs
+++ b/tests/ui/privacy/pub-priv-dep/pub-priv1.rs
@@ -32,6 +32,33 @@ pub struct PublicType {
     pub other_field: PubType, // Type from public dependency - this is fine
 }
 
+pub struct PublicTuple(
+    pub OtherType,
+    //~^ ERROR type `OtherType` from private dependency 'priv_dep' in public interface [exported_private_dependencies]
+    OtherType,
+    pub PubType,
+);
+
+pub enum PublicEnum {
+    OtherType,
+    ActualOtherType(OtherType, PubType),
+    //~^ ERROR type `OtherType` from private dependency 'priv_dep' in public interface [exported_private_dependencies]
+    ActualOtherTypeStruct {
+        field: OtherType,
+        //~^ ERROR type `OtherType` from private dependency 'priv_dep' in public interface [exported_private_dependencies]
+        other_field: PubType,
+    },
+}
+
+pub struct PublicGenericType<T, U>(pub T, U);
+pub type ReexportedPublicGeneric = PublicGenericType<OtherType, ()>;
+//~^ ERROR type `OtherType` from private dependency 'priv_dep' in public interface
+pub type ReexportedPrivateGeneric = PublicGenericType<(), OtherType>;
+//~^ ERROR type `OtherType` from private dependency 'priv_dep' in public interface
+
+pub struct PublicGenericBoundedType<T: OtherTrait>(T);
+//~^ ERROR trait `OtherTrait` from private dependency 'priv_dep' in public interface
+
 impl PublicType {
     pub fn pub_fn_param(param: OtherType) {}
     //~^ ERROR type `OtherType` from private dependency 'priv_dep' in public interface
@@ -46,9 +73,15 @@ pub trait MyPubTrait {
     type Foo: OtherTrait;
     //~^ ERROR trait `OtherTrait` from private dependency 'priv_dep' in public interface
 
-    fn required() -> impl OtherTrait;
+    fn required_impl_trait() -> impl OtherTrait;
 
-    fn provided() -> impl OtherTrait { OtherType }
+    fn provided_impl_trait() -> impl OtherTrait { OtherType }
+
+    fn required_concrete() -> OtherType;
+    //~^ ERROR type `OtherType` from private dependency 'priv_dep' in public interface
+
+    fn provided_concrete() -> OtherType { OtherType }
+    //~^ ERROR type `OtherType` from private dependency 'priv_dep' in public interface
 }
 
 pub trait WithSuperTrait: OtherTrait {}
@@ -67,6 +100,12 @@ impl PubLocalTraitWithAssoc for PrivateAssoc {
 pub fn in_bounds<T: OtherTrait>(x: T) { unimplemented!() }
 //~^ ERROR trait `OtherTrait` from private dependency 'priv_dep' in public interface
 
+pub fn private_return_impl_trait() -> impl OtherTrait { OtherType }
+//~^ ERROR trait `OtherTrait` from private dependency 'priv_dep' in public interface
+
+pub fn private_return() -> OtherType { OtherType }
+//~^ ERROR type `OtherType` from private dependency 'priv_dep' in public interface
+
 pub fn private_in_generic() -> std::num::Saturating<OtherType> { unimplemented!() }
 //~^ ERROR type `OtherType` from private dependency 'priv_dep' in public interface
 
@@ -79,6 +118,9 @@ pub const CONST: OtherType = OtherType;
 pub type Alias = OtherType;
 //~^ ERROR type `OtherType` from private dependency 'priv_dep' in public interface
 
+pub type AliasOfAlias = priv_dep::PubPub;
+//~^ ERROR type `OtherType` from private dependency 'priv_dep' in public interface
+
 pub struct PublicWithPrivateImpl;
 
 impl OtherTrait for PublicWithPrivateImpl {}
@@ -89,6 +131,22 @@ pub trait PubTraitOnPrivate {}
 impl PubTraitOnPrivate for OtherType {}
 //~^ ERROR type `OtherType` from private dependency 'priv_dep' in public interface
 //~| ERROR type `OtherType` from private dependency 'priv_dep' in public interface
+
+pub struct PublicWithStdImpl;
+
+impl From<OtherType> for PublicWithStdImpl {
+//~^ ERROR type `OtherType` from private dependency 'priv_dep' in public interface
+    fn from(val: OtherType) -> Self { Self }
+    //~^ ERROR type `OtherType` from private dependency 'priv_dep' in public interface
+}
+
+impl From<PublicWithStdImpl> for OtherType {
+    //~^ ERROR type `OtherType` from private dependency 'priv_dep' in public interface
+    //~| ERROR type `OtherType` from private dependency 'priv_dep' in public interface
+    fn from(val: PublicWithStdImpl) -> Self { Self }
+    //~^ ERROR type `OtherType` from private dependency 'priv_dep' in public interface
+    //~| ERROR type `OtherType` from private dependency 'priv_dep' in public interface
+}
 
 pub struct AllowedPrivType {
     #[allow(exported_private_dependencies)]
@@ -106,5 +164,14 @@ pub use pm::pm_attr;
 
 pub use priv_dep::E::V1;
 //~^ ERROR variant `V1` from private dependency 'priv_dep' is re-exported
+
+pub use priv_dep::Unit;
+//~^ ERROR type alias `Unit` from private dependency 'priv_dep' is re-exported
+pub use priv_dep::PubPub;
+//~^ ERROR type alias `PubPub` from private dependency 'priv_dep' is re-exported
+pub use priv_dep::PubPriv;
+//~^ ERROR type alias `PubPriv` from private dependency 'priv_dep' is re-exported
+pub use priv_dep::OtherType as Renamed;
+//~^ ERROR struct `Renamed` from private dependency 'priv_dep' is re-exported
 
 fn main() {}

--- a/tests/ui/privacy/pub-priv-dep/pub-priv1.stderr
+++ b/tests/ui/privacy/pub-priv-dep/pub-priv1.stderr
@@ -11,34 +11,58 @@ LL | #![deny(exported_private_dependencies)]
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 error: macro `m` from private dependency 'priv_dep' is re-exported
-  --> $DIR/pub-priv1.rs:98:9
+  --> $DIR/pub-priv1.rs:156:9
    |
 LL | pub use priv_dep::m;
    |         ^^^^^^^^^^^
 
 error: macro `fn_like` from private dependency 'pm' is re-exported
-  --> $DIR/pub-priv1.rs:100:9
+  --> $DIR/pub-priv1.rs:158:9
    |
 LL | pub use pm::fn_like;
    |         ^^^^^^^^^^^
 
 error: derive macro `PmDerive` from private dependency 'pm' is re-exported
-  --> $DIR/pub-priv1.rs:102:9
+  --> $DIR/pub-priv1.rs:160:9
    |
 LL | pub use pm::PmDerive;
    |         ^^^^^^^^^^^^
 
 error: attribute macro `pm_attr` from private dependency 'pm' is re-exported
-  --> $DIR/pub-priv1.rs:104:9
+  --> $DIR/pub-priv1.rs:162:9
    |
 LL | pub use pm::pm_attr;
    |         ^^^^^^^^^^^
 
 error: variant `V1` from private dependency 'priv_dep' is re-exported
-  --> $DIR/pub-priv1.rs:107:9
+  --> $DIR/pub-priv1.rs:165:9
    |
 LL | pub use priv_dep::E::V1;
    |         ^^^^^^^^^^^^^^^
+
+error: type alias `Unit` from private dependency 'priv_dep' is re-exported
+  --> $DIR/pub-priv1.rs:168:9
+   |
+LL | pub use priv_dep::Unit;
+   |         ^^^^^^^^^^^^^^
+
+error: type alias `PubPub` from private dependency 'priv_dep' is re-exported
+  --> $DIR/pub-priv1.rs:170:9
+   |
+LL | pub use priv_dep::PubPub;
+   |         ^^^^^^^^^^^^^^^^
+
+error: type alias `PubPriv` from private dependency 'priv_dep' is re-exported
+  --> $DIR/pub-priv1.rs:172:9
+   |
+LL | pub use priv_dep::PubPriv;
+   |         ^^^^^^^^^^^^^^^^^
+
+error: struct `Renamed` from private dependency 'priv_dep' is re-exported
+  --> $DIR/pub-priv1.rs:174:9
+   |
+LL | pub use priv_dep::OtherType as Renamed;
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 error: type `OtherType` from private dependency 'priv_dep' in public interface
   --> $DIR/pub-priv1.rs:29:5
@@ -49,82 +73,188 @@ LL |     pub field: OtherType,
 error: type `OtherType` from private dependency 'priv_dep' in public interface
   --> $DIR/pub-priv1.rs:36:5
    |
+LL |     pub OtherType,
+   |     ^^^^^^^^^^^^^
+
+error: type `OtherType` from private dependency 'priv_dep' in public interface
+  --> $DIR/pub-priv1.rs:44:21
+   |
+LL |     ActualOtherType(OtherType, PubType),
+   |                     ^^^^^^^^^
+
+error: type `OtherType` from private dependency 'priv_dep' in public interface
+  --> $DIR/pub-priv1.rs:47:9
+   |
+LL |         field: OtherType,
+   |         ^^^^^^^^^^^^^^^^
+
+error: type `OtherType` from private dependency 'priv_dep' in public interface
+  --> $DIR/pub-priv1.rs:54:1
+   |
+LL | pub type ReexportedPublicGeneric = PublicGenericType<OtherType, ()>;
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: type `OtherType` from private dependency 'priv_dep' in public interface
+  --> $DIR/pub-priv1.rs:56:1
+   |
+LL | pub type ReexportedPrivateGeneric = PublicGenericType<(), OtherType>;
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: trait `OtherTrait` from private dependency 'priv_dep' in public interface
+  --> $DIR/pub-priv1.rs:59:1
+   |
+LL | pub struct PublicGenericBoundedType<T: OtherTrait>(T);
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: type `OtherType` from private dependency 'priv_dep' in public interface
+  --> $DIR/pub-priv1.rs:63:5
+   |
 LL |     pub fn pub_fn_param(param: OtherType) {}
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 error: type `OtherType` from private dependency 'priv_dep' in public interface
-  --> $DIR/pub-priv1.rs:39:5
+  --> $DIR/pub-priv1.rs:66:5
    |
 LL |     pub fn pub_fn_return() -> OtherType { OtherType }
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 error: trait `OtherTrait` from private dependency 'priv_dep' in public interface
-  --> $DIR/pub-priv1.rs:46:5
+  --> $DIR/pub-priv1.rs:73:5
    |
 LL |     type Foo: OtherTrait;
    |     ^^^^^^^^^^^^^^^^^^^^
 
+error: type `OtherType` from private dependency 'priv_dep' in public interface
+  --> $DIR/pub-priv1.rs:80:5
+   |
+LL |     fn required_concrete() -> OtherType;
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: type `OtherType` from private dependency 'priv_dep' in public interface
+  --> $DIR/pub-priv1.rs:83:5
+   |
+LL |     fn provided_concrete() -> OtherType { OtherType }
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
 error: trait `OtherTrait` from private dependency 'priv_dep' in public interface
-  --> $DIR/pub-priv1.rs:54:1
+  --> $DIR/pub-priv1.rs:87:1
    |
 LL | pub trait WithSuperTrait: OtherTrait {}
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 error: type `OtherType` from private dependency 'priv_dep' in public interface
-  --> $DIR/pub-priv1.rs:63:5
+  --> $DIR/pub-priv1.rs:96:5
    |
 LL |     type X = OtherType;
    |     ^^^^^^
 
 error: trait `OtherTrait` from private dependency 'priv_dep' in public interface
-  --> $DIR/pub-priv1.rs:67:1
+  --> $DIR/pub-priv1.rs:100:1
    |
 LL | pub fn in_bounds<T: OtherTrait>(x: T) { unimplemented!() }
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
+error: trait `OtherTrait` from private dependency 'priv_dep' in public interface
+  --> $DIR/pub-priv1.rs:103:1
+   |
+LL | pub fn private_return_impl_trait() -> impl OtherTrait { OtherType }
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
 error: type `OtherType` from private dependency 'priv_dep' in public interface
-  --> $DIR/pub-priv1.rs:70:1
+  --> $DIR/pub-priv1.rs:106:1
+   |
+LL | pub fn private_return() -> OtherType { OtherType }
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: type `OtherType` from private dependency 'priv_dep' in public interface
+  --> $DIR/pub-priv1.rs:109:1
    |
 LL | pub fn private_in_generic() -> std::num::Saturating<OtherType> { unimplemented!() }
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 error: type `OtherType` from private dependency 'priv_dep' in public interface
-  --> $DIR/pub-priv1.rs:73:1
+  --> $DIR/pub-priv1.rs:112:1
    |
 LL | pub static STATIC: OtherType = OtherType;
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 error: type `OtherType` from private dependency 'priv_dep' in public interface
-  --> $DIR/pub-priv1.rs:76:1
+  --> $DIR/pub-priv1.rs:115:1
    |
 LL | pub const CONST: OtherType = OtherType;
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 error: type `OtherType` from private dependency 'priv_dep' in public interface
-  --> $DIR/pub-priv1.rs:79:1
+  --> $DIR/pub-priv1.rs:118:1
    |
 LL | pub type Alias = OtherType;
    | ^^^^^^^^^^^^^^
 
+error: type `OtherType` from private dependency 'priv_dep' in public interface
+  --> $DIR/pub-priv1.rs:121:1
+   |
+LL | pub type AliasOfAlias = priv_dep::PubPub;
+   | ^^^^^^^^^^^^^^^^^^^^^
+
 error: trait `OtherTrait` from private dependency 'priv_dep' in public interface
-  --> $DIR/pub-priv1.rs:84:1
+  --> $DIR/pub-priv1.rs:126:1
    |
 LL | impl OtherTrait for PublicWithPrivateImpl {}
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 error: type `OtherType` from private dependency 'priv_dep' in public interface
-  --> $DIR/pub-priv1.rs:89:1
+  --> $DIR/pub-priv1.rs:131:1
    |
 LL | impl PubTraitOnPrivate for OtherType {}
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 error: type `OtherType` from private dependency 'priv_dep' in public interface
-  --> $DIR/pub-priv1.rs:89:1
+  --> $DIR/pub-priv1.rs:131:1
    |
 LL | impl PubTraitOnPrivate for OtherType {}
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
 
-error: aborting due to 20 previous errors
+error: type `OtherType` from private dependency 'priv_dep' in public interface
+  --> $DIR/pub-priv1.rs:137:1
+   |
+LL | impl From<OtherType> for PublicWithStdImpl {
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: type `OtherType` from private dependency 'priv_dep' in public interface
+  --> $DIR/pub-priv1.rs:139:5
+   |
+LL |     fn from(val: OtherType) -> Self { Self }
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: type `OtherType` from private dependency 'priv_dep' in public interface
+  --> $DIR/pub-priv1.rs:143:1
+   |
+LL | impl From<PublicWithStdImpl> for OtherType {
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: type `OtherType` from private dependency 'priv_dep' in public interface
+  --> $DIR/pub-priv1.rs:143:1
+   |
+LL | impl From<PublicWithStdImpl> for OtherType {
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+
+error: type `OtherType` from private dependency 'priv_dep' in public interface
+  --> $DIR/pub-priv1.rs:146:5
+   |
+LL |     fn from(val: PublicWithStdImpl) -> Self { Self }
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: type `OtherType` from private dependency 'priv_dep' in public interface
+  --> $DIR/pub-priv1.rs:146:5
+   |
+LL |     fn from(val: PublicWithStdImpl) -> Self { Self }
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+
+error: aborting due to 41 previous errors
 

